### PR TITLE
perf(worker): fetch next job on failure

### PR DESCRIPTION
**Pull request by roggervalf**

### CODE REVIEW

**Review by manast — COMMENTED** [review]
This is a small but a breaking change 😅

File: src/classes/worker.ts [hunk 1]
```diff
@@ -715,8 +715,14 @@ export class Worker<
             return;
           }
 
-          await job.moveToFailed(err, token);
+          const failed = await job.moveToFailed(err, token, true);
           this.emit('failed', job, err, 'active');
+
+          if (failed) {
+            const [jobData, jobId, limitUntil, delayUntil] = failed || [];
```
Comment (manast) [inline src/classes/worker.ts:722]: || [] is redundant as it is already checking if failed is truthy

Comment (roggervalf) [reply src/classes/worker.ts]: I refactored to be similar to completed handler

File: src/classes/worker.ts [hunk 2]
```diff
@@ -766,8 +766,14 @@ will never work with more accuracy than 1ms. */
             return;
           }
 
-          await job.moveToFailed(err, token);
+          const failed = await job.moveToFailed(err, token, true);
```
Comment (manast) [inline src/classes/worker.ts:769]: I think "failed" here is misleading, as this includes the next job to process. Should be called probably "result" or similar.

File: src/classes/worker.ts [hunk 3]
```diff
@@ -786,7 +792,7 @@ will never work with more accuracy than 1ms. */
       const result = await this.callProcessJob(job, token);
       return await handleCompleted(result);
     } catch (err) {
-      return handleFailed(<Error>err);
+      return await handleFailed(<Error>err);
```
Comment (manast) [inline src/classes/worker.ts:795]: await here will imply an unhandled exception if handleFailed throws an exception, is this the intent?

Comment (roggervalf) [reply src/classes/worker.ts]: it was placed when testing, removing it now

File: src/classes/worker.ts (hunk 2 above)
Comment (roggervalf) [reply src/classes/worker.ts]: sounds good to me

**Review by manast — APPROVED** [review]
LGTM.  I see this PR more from a consistency side, as failed jobs are a deviation of the normal functioning of the queue and therefore it is not so important to optimize for this path, however having the same code for both completed and failed jobs is beneficial in my opinion.

**Pull request merged**